zug/util.hpp

```diff
@@ -52,48 +57,63 @@ ZUG_INLINE_CONSTEXPR struct identity__t
 
 namespace detail {
 
-template <class F, class G>
+template <typename Args, typename Fn1, typename Fn2>
+constexpr decltype(auto) apply_composition(Args&& args, Fn1&& fn1, Fn2&& fn2)
+{
+    return compat::invoke(
+        std::forward<Fn1>(fn1),
+        compat::apply(std::forward<Fn2>(fn2), std::forward<Args>(args)));
+}
+
+template <typename Args, typename Fn, typename... Fns, typename... T>
+constexpr decltype(auto) apply_composition(Args&& args, Fn&& fn, Fns&&... fns)
+{
+    return compat::invoke(
+        std::forward<Fn>(fn),
+        apply_composition(std::forward<Args>(args), std::forward<Fns>(fns)...));
+}
+
+template <typename Fns, typename Args, size_t... Is>
+constexpr decltype(auto)
+apply_composition(Fns&& fns, Args&& args, std::index_sequence<Is...>)
+{
+    return apply_composition(std::forward<Args>(args),
+                             std::get<Is>(std::forward<Fns>(fns))...);
+}
+
+template <typename Fns, typename Args>
+constexpr decltype(auto) apply_composition(Fns&& fns, Args&& args)
+{
+    return apply_composition(
+        std::forward<Fns>(fns),
+        std::forward<Args>(args),
+        std::make_index_sequence<std::tuple_size<std::decay_t<Fns>>::value>{});
+}
+
+template <typename Fn, typename... Fns>
 struct composed
```

Comment:
This will be renamed to `composable` or something in a future step. I would then create a helper function `make_composable` which would then be the basis for the `operator|`. That will then require at this point a function with signature `decltype(auto) apply_composition(Fn&&, Args&&)` (i.e., one function). This isn't currently possible going directly through `comp` (since the singular function just forwards).